### ISSUE TITLE
CollectionAssert.AreEqual should always compare elements, ignoring IEquatable if implemented

### ISSUE THREAD

**Issue by JeffAshton** [nunit/nunit#2097]
I came across this issue when comparing collections of type System.Collections.Immutable.ImmutableArray<T>.  This works correctly in NUnit 2, but now fails in NUnit 3.  The reason is that ImmutableArray implements IEquatable<T> under the covers and checks that the underlying arrays are the same reference.

https://github.com/dotnet/corefx/blob/master/src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.Minimal.cs#L317

CollectionAssert.AreEqual claims to check the items, but with the reuse of NUnitEqualityComparer.  This likely worked before this change that put comparing IEquatable<T> before IEnumerable<T>, but then comparing objects which also happen to implement IEnumerable<T> would have been broken.

https://github.com/nunit/nunit/commit/7c7487036accf1d530f4338fbd74d62dc9175993#diff-ebfb835776d7283aa42e5db9b6403200

When I'm comparing and Enumerable<T>, I shouldn't know about Equals(object) or IEquatable<T>, or IDictionary<TKey,TValue>, etc.  I'm giving an stream of items, and should only treat it as such.  That fact that CollectionAssert takes IEnumerable, but then does casting under the covers is a huge smell.

When I'm comparing Object's, I shouldn't know about enumerables.  For example, consider an implementation for [System.Collection.IGrouping<TKey,TElement>](https://msdn.microsoft.com/en-us/library/bb344977(v=vs.110).aspx).  To be equal, both the key and items have to match, but the implementation for NUnitEqualityComparer would disregard the Key altogether because they're an Enumerable<T>.

**Comment by JeffAshton** [comment]
In reviewing the implementation for [NUnitEqualityComparer.AreEqual](https://github.com/nunit/nunit/blob/757feb7a721ee8ef21b20e76ada61d00834c3bc7/src/NUnitFramework/framework/Constraints/NUnitEqualityComparer.cs#L132), I can also say that anything which implements IDictionary is problematic to the same as IEnumerable<T>.

It is only safe to do this sort of optimization for well known:

* structs
* sealed classes
* specific instance of a type, but those which inherit ( for example if type is Dictionary<TKey,TValue>, but it isn't safe for:

```
public class ExtendedDictionary<TKey,TValue> : Dictionary<TKey, TValue > {
    public int ExtendedProperty { get; }
}
```

**Comment by CharliePoole** [comment]
I'll review and comment on your sample tests but I have one general question... does your new title intend to say that CollectionAssert.AreEqual and Assert.AreEqual have different behaviors? This would be a bigger surprise to me than anything else because they should be (and once were) simply adapters to the same underlying code.

**Comment by CharliePoole** [comment]
I'll have to explain a bit of history first... trying not to be too verbose...

1. CollectionAssert was written many years ago to provide users of MSTest an equivalent to their CollectionAssert. It originally supported only collections, hence had only tests of collections. Originally, it had it's own code, just as Assert.AreEqual did, but when we implemented constraint-based assertions, both of them were rewritten to simply frontend the underlying comparison. That's when we changed the signatures of CollectionAssert to allow use of IEnumerable - because it was "free". We didn't add more tests, which was obviously an error.

2. NUnit - like it or not - has had it's own definition of what "AreEqual" means since the beginning, i.e. around 2000. I could write a pretty substantial book chapter on how that definition differs from mathematical equality and an appendix about why it does, but I promise not to. 😄 

3. At some point in the last year or two, folks asked us through the bug process to give precedence to any implementation of IEquatable when an object implemented IEnumerable. We discussed it a lot before doing it. The logic was mainly this: If the user implemented IEquatable for a class, it was obviously intended to be used.

Enough history... back to your issue...

Using IEquatable appears as a bug to you, but not using it was already recognized as a bug in the past. That's a bit of a conflict. I definitely think we have to use it when there is a user implementation. In this case it's a MIcrosoft implementation so maybe that's different. Let's hear what you think and what others think about the issue.

CollectionAssert probably needs its documentation checked. Frankly it's a bit of a legacy thing, which I've tried to dump in the past, since it does nothing that Assert doesn't do, often better. But some people seem attached to it, while others are confused and think they have to use it for collections.

I'm not sure I completely understand the last part of your comment, starting with "When I'm comparing an enumerable..." IF what you are saying is that NUnit should act differently depending on how a value is passed to it - e.g. (object)actual versus (IEnumerable)actual - I agree in principal but I'm afraid it's not going to happen soon. Internally, everything passed to NUnit ends up as an object. That's why NUnitEqualityComparer has all that awful code that tries to figure out the underlying types.

Bottom line, the equality comparison is doing what we at least thought it ought to do when we developed it. Any arguments about changing that have to take into account approximately 17 years of backward compatibility (but only a few years wrt IEquatable).

I'm keeping this open

**Comment by JeffAshton** [comment]
Definitely can appreciate history and it's affects on interfaces and frameworks.  :-)  I'm currently migrating my company from NUnit 2 to 3, and hence why I'm coming across these issues.   We've always used CollectionAssert for comparing enumerables and Assert for comparing objects.   Definitely a time and place for doing both.  When comparing objects,  I really like having IEquatable<T> as it gives a way to customize equality for types where we've overriden the == operator.  

Our portfolio of tests undercovered the System.Collections.Immutable issue, and I spotted the IGrouping<T> issue when investigating the other.   That one might actually have me more scared, as it's checking less than what it should for a subset of objects.  The immediate array issue just feels like a loss in capability.

One thing my coworker mentioned over lunch one day was that, "he was surprised that NUnit hadn't separated out the Assertion portion of the library".  Like how mocha doesn't care whether you use Assert or Expect libraries in the node community.  NUnit is obviously extensible in the same way, but having the out of the box Assert pushes this legacy on developers unconsciously.

**Comment by JeffAshton** [comment]
@CharliePoole :  I think we start by fixing CollectionAssert.AreEqual to go back to treating the values as IEnumerable.  Since IEnumerable doesn't extend IEquatable<T>, we shouldn't be boxing to an object then casting back.  ICollection is really just a glorified IEnumerable + Count, and would have suffered from the same issue.  It's important to realize that CollectionAssert and Assert are geared towards different use cases.

CollectionAssert should always treat the inputs as IEnumerable, where as Assert should always treat the inputs as objects and adhere to standard object comparison rules:

* IEqualityComparer&lt;T&gt; if provided
* IEqualityComparer if provided
* IEquatable&lt;T&gt; if implemented
* == operator if implemented ( maybe )
* Object.Equals last resort

**Comment by CharliePoole** [comment]
@JeffAshton The distinction between CollectionAssert and Assert as you outline it makes sense if we were designing both from scratch. That is, it's not the only sensible approach, but it's one of them.

However, we aren't designing them from scratch. There is a great deal of history behind the two of them working in exactly the same way, with any differences in the past having been treated as a bug. In addition, it's important to realize that CollectionAssert is seen (by us) as legacy, so we don't generally change it or add new features. The same is true for Assert.AreEqual and other pre-constraint assertions. They don't get new features in themselves although they sometimes benefit (or maybe suffer) from improvements in the back end.

Anyway, there may be something we can do here but redefining CollectionAssert is probably not going to fly... I'd say we either have to document it better or deprecate and eventually remove it.

**Comment by CharliePoole** [comment]
@JeffAshton ON separating assertions portion, see #502 

I was a big fan of this for a while, but there wasn't a lot of interest on the part of others and my own plate is pretty full so it was closed.

**Comment by CharliePoole** [comment]
@JeffAshton I'm still a little confused as to whether your issue title really means to imply that CollectionAssert.AreEqual is working differently from Assert.AreEqual. They should both be giving the same result for any given case. If you can clarify that it will be easier to move on with figuring out how to resolve the issue.

**Comment by jnm2** [comment]
> CollectionAssert should always treat the inputs as IEnumerable, where as Assert should always treat the inputs as objects and adhere to standard object comparison rules:
> 
> * IEqualityComparer&lt;T&gt; if provided
> * IEqualityComparer if provided
> * IEquatable&lt;T&gt; if implemented
> * == operator if implemented ( maybe )
> * Object.Equals last resort

I'm pretty sure the standard .NET equality comparison is `(providedComparer ?? EqualityComparer<T>Default).Equals(x, y)`. That takes care of the entire list, and the standard for .NET equality is to not check custom `==` or `!=` implementations.

**Comment by JeffAshton** [comment]
@jnm2 :  My point about Object equality rules was more to do with the bonus rules in [NUnitEqualityComparer](https://github.com/nunit/nunit/blob/757feb7a721ee8ef21b20e76ada61d00834c3bc7/src/NUnitFramework/framework/Constraints/NUnitEqualityComparer.cs#L240), specifically around IEnumerable.  Just because two objects are IEnumerable, doesn't mean they are equal if all items are equal, as demonstrated by [IGrouping&lt;T&gt;](https://msdn.microsoft.com/en-us/library/bb344977(v=vs.110).aspx) which is an IEnumerable + Key property.

The following simple [test case](https://github.com/nunit/nunit/pull/2098/files#diff-9545a4306db2da76400642e233a1c600R537) doesn't work in NUnit, since IGrouping&lt;T&gt; is compared as an IEnumerable&lt;T&gt; but shouldn't be.

```
IGrouping<string, int> objA = new SimpleGrouping<string, int>("abc", 1, 2);
IGrouping<string, int> objB = new SimpleGrouping<string, int>("xyz", 1, 2);
Assert.AreNotEqual(objA, objB, "Should not be equal because the keys don't match");
```

This is dangerous for consumers of NUnit, as no one would expect this behavior, and bugs could easily escape.

**Comment by JeffAshton** [comment]
@CharliePoole:  Really it comes down to the fact that I see value in what the CollectionAssert interface offers if it works as documented, and have used it as such for years.  That is treating the inputs as IEnumerable and comparing the individual values, irrelevant of backing Enumerable implementation.  The power of IEnumerable is that elements can come from anywhere.

I reported these 2 separate issues together only because they are related.  One can't be fixed by adjusting order within NUnitEqualityComparer, without breaking the other.   Comparing collections could use the NUnitEqualityComparer for comparing individual items, but ultimately the IEnumerable logic inside NUnitEqualityComparer should be pulled out, for reasons described by IGrouping&lt;TKey,TElement&gt;.  This violation of object inheritance is actually pretty dangerous if not caught by static analysis.

**Comment by jnm2** [comment]
I would definitely feel better knowing that objects are only treated as the sum of their values if they do not implement `IEquatable<>`.

**Comment by jnm2** [comment]
However, LINQ's backing class for `IGrouping` does not implement `IEquatable` so we'd still be in the same boat.

**Comment by jnm2** [comment]
The problem is that the way you compare enumerables in NUnit is `Is.EqualTo`, the same method used for object equality. I've always thought that was odd. I was looking for a method named something like `SameItems`.

**Comment by JeffAshton** [comment]
@jnm2, that's kind of what I'm getting at, and hence why I see value in the old CollectionAssert as defined, maybe just with a new name.  SameItems would mean same references (i.e. Object.ReferenceEquals).  We would want EqualItems as well.

**Comment by jnm2** [comment]
My late and irrelevant opinion is that `Is.EqualTo` should mean "equality as the object defines it." Almost no enumerables override `object.Equals`, and by that they define themselves as having reference equality. I would expect `Is.EqualTo` to respect that decision, and `Has.EqualItems` and `Has.SameItems` to iterate.

However, there is absolutely no way to change it in NUnit 3. Everyone's tests would break.

**Comment by CharliePoole** [comment]
@JeffAshton @jnm2 

It may well be that `Assert.AreEqual`, `CollectionAssert.AreEqual` and `EqualConstraint` __should__ have been designed to work in a different way - they certainly __could__ have.

But those ships sailed a long time ago, and not just with the release of NUnit 3. The decision that NUnit's definition for AreEqual would be different from what the object defines was made around the year 2000. CollectionAssert was designed in 2006. The constraint system, including EqualConstraint, was created in 2007. Folks have been using them heavily for all this time and I don't think we want to pull the rug out from under them.

That's not to say we can't change anything, but we have to be extremely careful in changing something that has worked in the same way for so many years. We made some breaking changes when we went to NUnit 3, but a fundamental change in the meaning of NUnit equality is much bigger than anything we changed then and wasn't something we even considered.

There's a certain logic to saying that CollectionAssert should always treat the arguments as collections. We didn't design it that way, but it's certainly what the name suggestss. We might be able to make that happen with only minor compatibility issues because  the exceptions were only introduced recently.

However, making Assert work different from how it is designed is a much bigger deal as far as compatibility goes. I also don't think it's the right thing to do, in spite of @jnm2's logic. Equality in a test is not the same as equality in math or in a programming language. We've had that discussion repeatedly over the years and have always opted for the definition that works pragmatically in a testing environment over a more "pure" approach. See, for example, how we handle NANs. 

I'm not criticizing you guys for wanting a different definition. There are plenty of things in NUnit that I would do differently in retrospect but which I live with because support for users trumps my notions of what makes a good design. That's the nature of development when it orients toward the user.

I'd like to make this issue about CollectionAssert. I think the issue with how EqualConstraint should work with certain arguments is a different one @JeffAshton even though you see some underlying causes that are the same. The main distinction is that we may need to make separate decisions about them.

For this issue, if it's going ahead, we need to figure out the minimal change to CollectionAssert's semantics that will will both resolve the problem and keep users existing tests working correctly to the greatest extent possible.

A useful bit of info for us, since CollectionAssert claims to work like the MSTest class of the same name, is how they handle the same situation. Does anyone know?

I'm marking this for both design and documentation, since it has aspects of both. At the moment, the design we need is of the "How should this work?" kind. Specifically...
1. Should the we work as @JeffAshton asks when comparing two ImmutableArrays?
2. What about one ImmutableArray compared to some other collection or enumerable?
3. As a stronger option, should CollectionAssert __generally__ ignore the implementation of IEquatable and only compare contents.

Let's try to focus on those questions please.

**Comment by JeffAshton** [comment]
Sounds good @CharliePoole.  I wrote some tests with MSTest to compare who it handles CollectionAssert and ImmutableArray.  It appears to compare the elements of the enumerable as per the documentation.

https://github.com/JeffAshton/MSTestsVerification/blob/master/MSTestsVerification/CollectionAssertTests.cs

All of these tests past as they are written.  I wrote tests with ImmutableArray, but as well with a custom collection which implements both Object.Equals & IEquatable<T>.Equals.  This implementation simply throws a NotSupportedException to demonstrate that it doesn't care about the underlying implementation.

https://github.com/JeffAshton/MSTestsVerification/blob/master/MSTestsVerification/SimpleCollection.cs

**Comment by JeffAshton** [comment]
Now to answer your questions @CharliePoole, from a purest standpoint, If CollectionAssert takes either an IEnumerable or an ICollection, the implementing library must treat it as such.  Since neither of these extend IEquatable&lt;T&gt;, then I can't treat it as such.  One thing I preach is that casting is code smell, and should generally be avoided.

I'm actually a fan of NUnit's change from ICollection to IEnumerable for this use case.  While comparing the Count can be a nice optimization internally, being able to compare any enumeration of elements is super powerful.  For example most things returned from Linq are an IEnumerable&lt;T&gt;.

**Comment by JeffAshton** [comment]
I also ran the following test against previous versions of nunit.

```
[Test]
public void Test() {
	ImmutableArray<int> x = ImmutableArray.Create<int>( 1, 2, 3 );
	ImmutableArray<int> y = ImmutableArray.Create<int>( 1, 2, 3 );
	CollectionAssert.AreEqual( x, y );
}
```

"3.0.0" - Passed
"3.2.0" - Passed
"3.2.1" - Passed
"3.4.0" - Failed

The breaking change was introduced in 3.4.0 as a result of https://github.com/nunit/nunit/issues/1439.  This issue does make sense in terms of Assert.AreEqual, since this takes objects.

I feel the root of this issue is that [CollectionItemsEqualConstraint](https://github.com/nunit/nunit/blob/master/src/NUnitFramework/framework/Constraints/CollectionItemsEqualConstraint.cs) delegates enumeration to NUnitEqualityComparer .  If CollectionItemsEqualConstraint enumerated the elements and then compared them using NUnitEqualityComparer, then everything would be fine.

**Comment by JeffAshton** [comment]
An interesting follow up to my last comment, CollectionItemsEqualConstraint is the only collection constraint that doesn't enumerate.  Most of the collection constraints delegate to [CollectionTally](https://github.com/nunit/nunit/blob/master/src/NUnitFramework/framework/Constraints/CollectionTally.cs) which enumerates, but the ones which do not [CollectionContainsConstraint](https://github.com/nunit/nunit/blob/master/src/NUnitFramework/framework/Constraints/CollectionContainsConstraint.cs) and [CollectionOrderedConstraint](https://github.com/nunit/nunit/blob/master/src/NUnitFramework/framework/Constraints/CollectionOrderedConstraint.cs) also enumerate.

**Comment by CharliePoole** [comment]
@JeffAshton I appreciate your enthusiasm. 😄 However, I think the true "root of the issue" is that we (the NUnit team) expected `CollectionAssert.AreEqual` to work __exactly__ the same as `Assert.AreEqual`. You, on the other hand, expected it to act differently and would like us to change our concept for how it works. I'm waiting for some input from other members of the @nunit/core-team and @nunit/framework-team so we can decide.

Maybe it will simplify matters if I make a proposal for everyone to review...

I propose that `CollectionAssert` should __never__ treat it's arguments as anything other than a collection. Right now, that means ignoring any implementation of IEquatable, but there could be other things in the future. Note that I'm using lower case "collection" here to include any IEnumerable.

Reasons PRO:
1. The name `CollectionAssert` makes it sound like that's what it does, which is why @JeffAshton assumed it worked that way.
2. As far as we can tell, that's how mstest's `CollectionAssert` works. We wrote our own version in order to give comfort to users who were accustomed to the mstest syntax.

Reasons CON
1. This could make it more difficult (but not impossible) to eventually retire `CollectionAssert`, which has been a long-term goal of mine.
2. It requires modifications to the underlying constraints in a way avoids messing up the working of `Assert.AreEqual`.
3. It probably also requires exposing the option to compare as a collection at the constraint level, so that users can access it using `Assert.That`

Personally, I think the Pros outway the Cons. @rprouse @ChrisMaddock  What do you think?

**Comment by rprouse** [comment]
> I propose that CollectionAssert should never treat it's arguments as anything other than a collection. 

I agree with that. I don't think any of the cons are very serious. For the third, I don't see it as a con, just a task that we should consider. I would propose that we just add to the existing syntax to support collection equality.

**Comment by CharliePoole** [comment]
@rprouse I concur. 

We might consider re-purposing the `AsCollection` modifier to `EqualConstraint`. It's current use is to allow comparing two arrays with different shapes (e.g. a[5,10], b[10,5], c[50] ) based solely on the content. 
That doesn't really get much use and could easily be used to force using the collection comparison even in the presence of `IEquatable`. Then `CollectionAssert.AreEqual` could simply use the modifier.

**Comment by CharliePoole** [comment]
So... I'm convinced by @JeffAshton 's logic and am changing `CollectionAssert.AreEqual` to no longer work like `Assert.AreEqual`. `CollectionAssert` will ignore any Implementation of `IEquatable`. Overrides of Equals, without use of the interface, are already ignored by both classes.

In implementing this, I'm expanding the meaning of the existing modifier property `EqualConstraint.AsCollection` to indicate that any `IEquatable` implementation should be ignored. This will allow users of the more modern constraint syntax to access the same functionality.

@JeffAshton You combined this request with a second request that we stop using our definition of enumerable equality with `Assert.AreEqual` - if I understood you correctly, since we didn't dig into that one. If you want to pursue this, you can create a new issue. However, I'll say in advance that we won't globally remove our treatment of enumerables from NUnit's definition of equality because too many people count on it. We might be able to do some fine tuning around the handling of dictionaries, however.

**Comment by JeffAshton** [comment]
Thanks @CharliePoole for addressing this.

**Comment by CharliePoole** [comment]
@nunit/framework-team Can I get a review of this please?
